Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/Update.hs

```diff
@@ -14,58 +14,134 @@
+      >>= listToMaybe
+      >>= (.fulfillmentState)
+      >>= (.fulfillmentStateDescriptor)
+      >>= (.descriptorCode)
+      & fromMaybeM (InvalidRequest "Event type is not present in OnUpdateReq.")
+  case eventType of
+    "PAYMENT_COMPLETED" -> parsePaymentCompletedEvent req
+    "EDIT_LOCATION" -> parseEditLocationEvent req
+    _ -> throwError (InvalidRequest "Invalid event type")
 
-mkPaymentMethodInfo :: Update.Payment -> DMPM.PaymentMethodInfo
-mkPaymentMethodInfo Update.Payment {..} =
-  DMPM.PaymentMethodInfo
-    { collectedBy = Common.castPaymentCollector collected_by,
-      paymentType = Common.castPaymentType _type,
-      paymentInstrument = Common.castPaymentInstrument instrument
-    }
+parsePaymentCompletedEvent :: (MonadFlow m) => Spec.UpdateReqMessage -> m DUpdate.DUpdateReq
+parsePaymentCompletedEvent req = do
+  rideId <- req.updateReqMessageOrder.orderFulfillments >>= listToMaybe >>= (.fulfillmentId) & fromMaybeM (InvalidRequest "RideId not present")
+  headPayment <- req.updateReqMessageOrder.orderPayments >>= listToMaybe & fromMaybeM (InvalidRequest "Payment not present")
+  orderId <- req.updateReqMessageOrder.orderId & fromMaybeM (InvalidRequest "OrderId not present")
+  paymentMethodInfo <- mkPaymentMethodInfo headPayment
+  return $
+    DUpdate.PaymentCompletedReq
+      { bookingId = Id orderId,
+        rideId = Id rideId,
+        paymentStatus = castPaymentStatus headPayment.paymentStatus,
+        paymentMethodInfo = paymentMethodInfo
+      }
 
-castPaymentStatus :: Update.PaymentStatus -> DUpdate.PaymentStatus
-castPaymentStatus Update.PAID = DUpdate.PAID
-castPaymentStatus Update.NOT_PAID = DUpdate.NOT_PAID
+parseEditLocationEvent :: (MonadFlow m) => Spec.UpdateReqMessage -> m DUpdate.DUpdateReq
+parseEditLocationEvent req = do
+  rideId <- req.updateReqMessageOrder.orderFulfillments >>= listToMaybe >>= (.fulfillmentId) & fromMaybeM (InvalidRequest "RideId not present")
+  headStops <- req.updateReqMessageOrder.orderFulfillments >>= listToMaybe & fromMaybeM (InvalidRequest "No Locations present")
+  bookingId <- (req.updateReqMessageOrder.orderId) & fromMaybeM (InvalidRequest "bookingId not present")
+  stops <- headStops.fulfillmentStops & fromMaybeM (InvalidRequest "No Locations present")
+  origin <- findLocationByStopType stops "START"
+  destination <- findLocationByStopType stops "END"
+  return $
+    DUpdate.EditLocationReq
+      { bookingId = Id bookingId,
+        rideId = Id rideId,
+        ..
+      }
+  where
+    findLocationByStopType :: MonadFlow m => [Spec.Stop] -> Text -> m (Maybe DL.Location)
+    findLocationByStopType stopsList stopType' = do
+      let res = find (\stop -> stop.stopType == Just stopType') stopsList
+      case res of
+        Just stop -> do
+          fromStopLocation <- fromMaybeM (InvalidRequest "Stop Locations not present") stop.stopLocation
+          stopLocation <- buildLocation fromStopLocation
+          return $ Just stopLocation
```

Comment:
Can this be replaced by a single line? Something like this

`stopLocation <- fromMaybeM (InvalidRequest "Stop Locations not present") $ stop.stopLocation >>= \location -> fmap  (buildLocation location) . Just`